pkg/storage/image.go

```diff
@@ -176,145 +178,106 @@ func (svc *imageService) makeRepoDigests(knownRepoDigests, tags []string, imageI
 	return imageDigest, repoDigests
 }
 
+func (svc *imageService) buildImageCacheItem(systemContext *types.SystemContext, ref types.ImageReference, image *storage.Image) (imageCacheItem, error) {
+	img, err := ref.NewImageSource(systemContext)
+	if err != nil {
+		return imageCacheItem{}, err
+	}
+	size := imageSize(img)
+	configDigest, err := imageConfigDigest(img, nil)
+	img.Close()
+	if err != nil {
+		return imageCacheItem{}, err
+	}
+	imageFull, err := ref.NewImage(systemContext)
+	if err != nil {
+		return imageCacheItem{}, err
+	}
+	defer imageFull.Close()
+	imageConfig, err := imageFull.OCIConfig()
+	if err != nil {
+		return imageCacheItem{}, err
+	}
+	return imageCacheItem{
+		user:         imageConfig.Config.User,
+		size:         size,
+		configDigest: configDigest,
+	}, nil
+}
+
+func (svc *imageService) appendCachedResult(systemContext *types.SystemContext, ref types.ImageReference, image *storage.Image, results []ImageResult, newImageCache imageCache) ([]ImageResult, error) {
+	var err error
+	svc.imageCacheLock.Lock()
+	cacheItem, ok := svc.imageCache[image.ID]
+	svc.imageCacheLock.Unlock()
+	if !ok {
+		cacheItem, err = svc.buildImageCacheItem(systemContext, ref, image)
+		if err != nil {
+			return results, err
+		}
+		if newImageCache == nil {
+			svc.imageCacheLock.Lock()
+			svc.imageCache[image.ID] = cacheItem
+			svc.imageCacheLock.Unlock()
+		} else {
+			newImageCache[image.ID] = cacheItem
+		}
+	} else if newImageCache != nil {
+		newImageCache[image.ID] = cacheItem
+	}
+
+	name, tags, digests := sortNamesByType(image.Names)
+	imageDigest, repoDigests := svc.makeRepoDigests(digests, tags, image.ID)
+	return append(results, ImageResult{
+		ID:           image.ID,
+		Name:         name,
+		RepoTags:     tags,
+		RepoDigests:  repoDigests,
+		Size:         cacheItem.size,
+		Digest:       imageDigest,
+		ConfigDigest: cacheItem.configDigest,
+		User:         cacheItem.user,
+	}), nil
+}
+
 func (svc *imageService) ListImages(systemContext *types.SystemContext, filter string) ([]ImageResult, error) {
-	results := []ImageResult{}
+	var results []ImageResult
 	if filter != "" {
+		// we never remove entries from cache unless unfiltered ListImages call is made. Is it safe?
```

Comment:
cache invalidation strategy which relies on how we are called doesn't feel right. Can it happen that this method is always called with filter only? What would be the consequence of having invalid cache entry?

Reply:
My expectation is that a cache entry that hasn't been pruned yet uses up memory, and map lookups can take incrementally longer.  We could have `UntagImage()` and `RemoveImage()` delete the entry, which would cover cases where deletion was done via CRI, but we'd still want to do a general pruning of the cache at some point, because local image storage can be modified by other processes.

Reply:
> because local image storage can be modified by other processes.

do I get it right, that `ListImages` is something like `docker ps` for kubelet? Can it cause problems a pod uses `IfNotPresent` image pull strategy and kubelet receives stale record from cache, that image is present?

Reply:
`ListImages` is what `crictl images` invokes, and is more directly comparable to the API that `docker images` calls.  The kubelet uses an `ImageStatus` call to check if an image is present before deciding whether or not it should call `PullImage`.

It would be a problem if we were returning stale cache entries to the kubelet.  We should only be consulting the cache for information about images which the storage library (either directly, or by way of the image library calling into it) says are still present in local storage.

Reply:
> We should only be consulting the cache for information about images which the storage library ... says are still present in local storage.

Is it enforced somehow now? Are we safe with current caching strategy?

Reply:
Yes, we currently read an item from the cache only when we have a corresponding `storage.Image` in hand, and that should only be retrievable if the image is present in local storage.